### ISSUE TITLE
Failover of NCCL Connection

### ISSUE THREAD

**Issue by yanminjia** [NVIDIA/nccl#998]
Specifically, during a collective communication such as Allreduce, if the current connection between 2 GPUs is broken because of such as network link failure, can we have a way to resume the Allreduce by changing to a different P2P connection if it is available between these 2 GPUs? 

For example, as attached, if the green link is broken, does the collective communication be completed via the PXN blue path?

[failover.pdf](https://github.com/NVIDIA/nccl/files/12640289/failover.pdf)

It wouldbe highly appreciated if any comment or suggestion. Many thanks.

**Comment by sjeaugey** [comment]
No, that's not what PXN is designed for, nor NCCL in general. If something fails (a link, NIC, GPU, CPU, .. anything), NCCL is going to fail, the application should either stop and restart from the latest checkpoint, or tear down the NCCL communicator and recreate a new one with systems that are still healthy (that requires a pretty advanced cluster fault detection and health check system to be coupled with the process launchger/manager).

**Comment by yanminjia** [comment]
Many thanks for your response, Syvain. Do you think if it is possible to add some kind of redundancy mechanism to NCCL by means of PXN or other better ways?

**Comment by sjeaugey** [comment]
It would be a huge amount of work, would only cover a specific type of failure, and would be very hard to maintain (automating random link loss and failover testing is not easy).

Keep in mind that PXN today is only implemented on the sender side, not on the receiver side.

**Comment by yanminjia** [comment]
Understood. Many thanks, Syvain.

**Comment by yanminjia** [comment]
Hello Syvain, as far as I know, in a use case of our partner, the application of the AI model learning generates a checkpoint per 3 hours because it takes much time and storage & computing resource to save the large amount of parameters. Therefore, if one collective communication such as Allreduce fails, the training will have to roll back to the checkpoint 3 hours ago in the worst case. It is certainly a waste. That is the reason why I think it does make sense to add redundancy mechanism to NCCL. 

If PXN is only available on the sender side, is there other way to copy the data to the GPU on the receiver side? In other word, if one RNIC is broken, can we transfer the data via other RNICs in case of RDMA? Thanks.

**Comment by sjeaugey** [comment]
It would be possible to go through the CPU (may be super slow though). Now, that won't happen transparently. You'd still need to tear down the NCCL communicator and recreate a new one (which would only use 7 ports instead of 8 one of the nodes.

**Comment by yanminjia** [comment]
My logic is that when the current connection between 2 inter-node GPUs is broken because of one RNIC failure, a backup/redundant connection is used by going with a different RNIC instead. So the discovered topology of GPUs, rings & trees, will not be changed. I'm not sure if that indicates that we don't need to tear down the NCCL communicator and recreate a new one. Otherwise, if recreate a new communicator, the ongoing collective communication would fail. And application of model training would have to rollback to the checkpoint.

**Comment by sjeaugey** [comment]
Well, it is a lot of changes.

There are many challenges as well -- just to explain a few:
 - When a send/recv operation fails, it's hard to know whether both sides have seen it fail the same way. You could have the message arrive and complete correctly on the receiver side, but the ack to the sender gets lost and generates an error. If the sender sends again (e.g. using a different NIC), the receiver may not have detected the problem and would not post the corresponding receive, leading to a hang.
 -  When you restart an operation using a different NIC, which NIC should you use? Another one at random? The closest one? Or should you now spread each message from this sender on all 7 other NICs?
 - If you want to use PXN, you can't do that at the network plugin level. You need to have the layer above first move the data elsewhere then send from the NIC local to that new GPU. Except, all buffers are pre-allocated/registered at init (or during the first send/recv for send/recv operations). So you'd also need to tell the CUDA kernels to use new buffers (new pointers), hence this functionality would have to be ingrained in pretty much every part of NCCL: the setup part, the CUDA kernels, and the network proxy code.

**Comment by yanminjia** [comment]
It looks really challenging. Let's think it differently to attack this issue. It would be an alternative solution if have a mechanism to roll back a failed collective communication to the original state. And then we can restart the collective communication by re-creating a new communicator. To be frank, I'm a beginner of NCCL code. My apology if I'm asking stupid questions.

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're [reviewing](https://github.com/NVIDIA/nccl/discussions/1761) older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
